pkg/runtime/dispatcher.go

```diff
@@ -0,0 +1,443 @@
+		nil,
+		func(call *dispatchInfo) []dispatchFn {
+			instCfg := make(map[string]proto.Message)
+			for _, inst := range call.instances {
+				instCfg[inst.Name] = inst.Params.(proto.Message)
+			}
+			return []dispatchFn{func(ctx context.Context) *result {
+				err := call.processor.ProcessReport(ctx, instCfg, requestBag, m.mapper, call.handler)
+				return &result{err: err, callinfo: call}
+			}}
+		},
+	)
+	return err
+}
+
+// Check dispatches to the set of adapters associated with the Check API method
+// Config validation ensures that things are consistent.
+// If they are not, we should continue as far as possible on the runtime path
+// before aborting. Returns an error if any of the adapters return an error.
+// If not the results are combined to a single CheckResult.
+// Dispatcher#Check.
+func (m *dispatcher) Check(ctx context.Context, requestBag attribute.Bag) (*adapter.CheckResult, error) {
+	cres, err := m.dispatch(ctx, requestBag, adptTmpl.TEMPLATE_VARIETY_CHECK,
+		nil,
+		func(call *dispatchInfo) []dispatchFn {
+			ra := make([]dispatchFn, 0, len(call.instances))
+			for _, inst := range call.instances {
+				ra = append(ra,
+					func(ctx context.Context) *result {
+						resp, err := call.processor.ProcessCheck(ctx, inst.Name,
+							inst.Params.(proto.Message),
+							requestBag, m.mapper,
+							call.handler)
+						return &result{err, &resp, call}
+					})
+			}
+			return ra
+		},
+	)
+
+	var res *adapter.CheckResult
+	if cres != nil {
+		res = cres.(*adapter.CheckResult)
+	}
+	return res, err
+}
+
+// Quota dispatches to the set of aspects associated with the Quota API method
+// Config validation ensures that things are consistent.
+// Quota calls are dispatched to at most one handler.
+// Dispatcher#Quota.
+func (m *dispatcher) Quota(ctx context.Context, requestBag attribute.Bag,
+	qma *aspect.QuotaMethodArgs) (*adapter.QuotaResult2, error) {
+	dispatched := false
+	qres, err := m.dispatch(ctx, requestBag, adptTmpl.TEMPLATE_VARIETY_QUOTA,
+		func(inst *cpb.Instance) bool {
+			return inst.Name == qma.Quota
+		},
+		func(call *dispatchInfo) []dispatchFn {
+			// assert len(call.instances) != 0 by construction.
+			inst := call.instances[0] // the 1st instance will be dispatched.
+			if dispatched {           // ensures only one call is dispatched.
+				glog.Warningf("Multiple dispatch: not dispatching %s to handler %s", inst.Name, call.handlerName)
+				return nil
+			}
+			dispatched = true
+			return []dispatchFn{
+				func(ctx context.Context) *result {
+					resp, err := call.processor.ProcessQuota(ctx, inst.Name,
+						inst.Params.(proto.Message), requestBag, m.mapper, call.handler,
+						adapter.QuotaRequestArgs{
+							DeduplicationID: qma.DeduplicationID,
+							QuotaAmount:     qma.Amount,
+							BestEffort:      qma.BestEffort,
+						})
+					return &result{err, &resp, call}
+				},
+			}
+		},
+	)
+	var res *adapter.QuotaResult2
+	if qres != nil {
```

Comment:
Is this nil check really necessary (here and in the other methods above)? Can't a nil value be casted?